fix: Fixed Maya submitter crashes with list attributes and special characters in file paths

sakshie95 opened the pull request:
Fixes: https://github.com/aws-deadline/deadline-cloud-for-maya/issues/348

### What was the problem/requirement? (What/Why)
The Maya submitter crashes and fails to launch when loading certain scene files. On looking more these were scene files containing Bifrost caches or file paths with special characters like brackets. The crashes occurred in two scenarios:

1. **Bifrost multi-attributes**: Maya's `getAttr()` returns lists for multi-attributes (like Bifrost cache paths) instead of strings, causing `AttributeError: 'list' object has no attribute 'replace'` when the code tried to call `.replace()` on the list.

2. **Regex special characters**: File paths containing regex special characters (`[`, `]`, `(`, `)`, etc.) caused `re.error: bad character range` when compiled as regex patterns in `findAllFilesForPattern()`.

### What was the solution? (How)
**Fix 1: Handle list attributes from Maya's getAttr()**
- Added `_flatten_and_validate_paths()` method to recursively flatten list values from `getAttr()` into strings
- Modified `_get_node_attr_paths()` to use the flattening method instead of assuming all values are strings

**Fix 2: Handle regex compilation errors**
- Added try/except block in `findAllFilesForPattern()` to catch `re.error` exceptions
- Falls back to literal file existence check when regex compilation fails

### What is the impact of this change?
- Users can now successfully load and submit Maya scenes with Bifrost caches
- Users can work with scenes containing file paths with special characters (brackets, parentheses, etc.)
- No breaking changes - existing functionality is preserved, only error handling is improved
- Better error messages for debugging when regex patterns fail

### How was this change tested?
**Unit tests**: Added 14 comprehensive unit tests covering both fixes:
- 7 tests for `_flatten_and_validate_paths()`: string attributes, list attributes, mixed types, empty values, nested lists, complex Bifrost scenarios
- 7 tests for `findAllFilesForPattern()` regex error handling: brackets, parentheses, braces, file exists/not exists cases, warning messages

All tests pass:
```
test/unit/deadline_submitter_for_maya/scripts/test_assets.py::TestFlattenAndValidatePaths - 7 passed test/unit/deadline_submitter_for_maya/scripts/test_utils.py::TestFindAllFilesForPatternRegexErrors - 7 passed
```

#### Integ test result
Submitter integ tests result:
```
test/integ/test_maya_submitters.py::TestSubmitters::test_scene_submitter[Minimal Maya Test] 
[gw0] [ 50%] PASSED test/integ/test_maya_submitters.py::TestSubmitters::test_scene_submitter[Minimal Maya Test] 
test/integ/test_maya_submitters.py::TestSubmitters::test_scene_submitter[Redshift Test] 
[gw0] [100%] PASSED test/integ/test_maya_submitters.py::TestSubmitters::test_scene_submitter[Redshift Test] 


======================================= slowest 5 durations ========================================
19.89s call     test/integ/test_maya_submitters.py::TestSubmitters::test_scene_submitter[Redshift Test]
3.77s setup    test/integ/test_maya_submitters.py::TestSubmitters::test_scene_submitter[Minimal Maya Test]
1.88s call     test/integ/test_maya_submitters.py::TestSubmitters::test_scene_submitter[Minimal Maya Test]
1.31s teardown test/integ/test_maya_submitters.py::TestSubmitters::test_scene_submitter[Redshift Test]
0.00s setup    test/integ/test_maya_submitters.py::TestSubmitters::test_scene_submitter[Redshift Test]
================================== 2 passed, 2 warnings in 30.44s ==================================

```

#### Installer test result
Installer built successfully and tested manually with the fixed code.

```
======================================= slowest 5 durations ========================================
11.65s setup    test/installer/test_installer.py::TestUserInstall::test_uninstall
11.59s setup    test/installer/test_installer.py::TestUserInstall::test_install
10.91s setup    test/installer/test_installer.py::TestLinuxAndMacOS::test_user_permissions
3.51s call     test/installer/test_installer.py::TestUserInstall::test_uninstall
2.50s call     test/installer/test_installer.py::test_default_location

```
#### Did you run the "Job Bundle Output Tests"? If not, why not? If so, paste the test results here.
N/A

### Was this change documented?
Relevant doc strings and comments were added

### Did you modify schema files?
- [ ] Yes
- [x] No

If yes, Did you bump the `integration_data_interface_version` in `src/deadline/maya_adaptor/MayaAdaptor/adaptor.py` and update the test constants?
See the "Schema Versioning" section in DEVELOPMENT.md for details.
   - [ ] Yes
   - [x] No

### Is this a breaking change?
No. This is a bug fix that maintains backward compatibility. All existing functionality is preserved - the changes only add error handling for edge cases that previously caused crashes.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

Commit:
fix: Fixed Maya submitter crashes with list attributes and special characters in file paths

Signed-off-by: Sakshi <sakshie@amazon.com>

leongdl commented on src/deadline/maya_submitter/utils.py:
```diff
@@ -64,11 +64,19 @@ def findAllFilesForPattern(pattern: str, frameNumber: int) -> list[str]:
+                if re.match(regex, f, flags=re.IGNORECASE) and os.path.isfile(os.path.join(dirname, f))
+            ]
+        except re.error as e:
+            # Handle paths with regex special characters (e.g., "/path/to/cache[1].bif")
```
I don't believe this, another `[` and `]` file name to regex bug in 1 week.

leongdl commented on src/deadline/maya_submitter/utils.py:
```diff
@@ -64,11 +64,19 @@ def findAllFilesForPattern(pattern: str, frameNumber: int) -> list[str]:
-            if re.match(regex, f, flags=re.IGNORECASE) and os.path.isfile(os.path.join(dirname, f))
-        ]
+        try:
+            regex = _patternToRegex(basename)
```
Can we escape the input string?

Godot's fix used a glob.escape function, is there similar for regex?

godobyte replied:
I think there is also `re.escape`.

sakshie95 replied:
Done, ty

godobyte commented on src/deadline/maya_submitter/assets.py:
```diff
@@ -267,6 +267,40 @@ def _get_arnold_texture_files(self) -> dict[str, Any]:
+                nested_paths = self._flatten_and_validate_paths(raw_path)
+                flattened_paths.extend(nested_paths)
+
+        return flattened_paths
```
Could we have an else clause to handle unexpected data types? That would improve the robustness of the function and make it easier to debug.

godobyte commented on src/deadline/maya_submitter/assets.py:
```diff
@@ -267,6 +267,40 @@ def _get_arnold_texture_files(self) -> dict[str, Any]:
+        for raw_path in raw_paths:
+            if raw_path is None:
+                # Skip None/empty values
+                continue
+            elif isinstance(raw_path, str):
+                # Single string path - add if not empty
+                if raw_path.strip():
+                    flattened_paths.append(raw_path)
+            elif isinstance(raw_path, list):
+                # List of paths - recursively flatten (handles nested lists if they occur)
+                nested_paths = self._flatten_and_validate_paths(raw_path)
+                flattened_paths.extend(nested_paths)
+
+        return flattened_paths
```
Could we have an else clause to handle unexpected data types? That would improve the robustness and make it easier to debug.

sakshie95 replied:
Done, ty

dependabot[bot] committed:
chore(deps): Update ruff requirement from ==0.13.* to ==0.14.* (#342)

Updates the requirements on [ruff](https://github.com/astral-sh/ruff) to permit the latest version.
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/0.13.0...0.14.0)

---
updated-dependencies:
- dependency-name: ruff
  dependency-version: 0.14.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): Bump actions/checkout from 5 to 6 (#349)

Bumps [actions/checkout](https://github.com/actions/checkout) from 5 to 6.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

sakshie95 committed:
Merge branch 'mainline' into fix/maya-submitter-crash
